plugins/cameraserver/src/main/java/edu/wpi/first/shuffleboard/plugin/cameraserver/source/CameraServerSource.java

```diff
@@ -0,0 +1,163 @@
+  private final MatOfByte conversionBuffer = new MatOfByte();
+
+  private final ExecutorService frameGrabberService = Executors.newSingleThreadExecutor(ThreadUtils::makeDaemonThread);
+  private final BooleanBinding enabled = active.and(connected);
+  private Future<?> frameFuture = null;
+
+  private CameraServerSource(String name) {
+    super(CameraServerDataType.INSTANCE);
+    setName(name);
+    ITable table = rootTable.getSubTable(name);
+    camera = new HttpCamera(name, removeCameraProtocols(table.getStringArray("streams", emptyStringArray)));
+    videoSink = new CvSink(name + "-videosink");
+    videoSink.setSource(camera);
+
+    enabled.addListener((__, was, is) -> {
+      // Disable the stream when not active or not connected to save on bandwidth
+      videoSink.setEnabled(is);
+      if (is) {
+        frameFuture = frameGrabberService.submit(this::grabForever);
+      } else if (frameFuture != null) {
+        frameFuture.cancel(true);
+      }
+    });
+
+    setActive(camera.getUrls().length > 0);
+    setConnected(true);
+
+    table.addTableListenerEx("streams", NetworkTableUtils.createListenerEx((source, key, value, flags) -> {
+      if (NetworkTableUtils.isDelete(flags) || !(value instanceof String[]) || ((String[]) value).length == 0) {
+        setActive(false);
+      } else {
+        String[] urls = removeCameraProtocols((String[]) value);
+        if (EqualityUtils.isDifferent(camera.getUrls(), urls)) {
+          camera.setUrls(urls);
+        }
+        setActive(true);
+      }
+    }), 0xFF);
+  }
+
+  /**
+   * Removes leading camera protocols from an array of stream URLs. These URLs are usually in the format
+   * {@code mjpg:http://...}, {@code ip:http://...}. This method will remove the leading {@code mjpg}.
+   *
+   * <p>This does not modify the existing array and returns a new array.
+   *
+   * @param streams an array of camera stream URLs to remove the
+   *
+   * @return the stream URLs without the leading camera protocols
+   */
+  private static String[] removeCameraProtocols(String... streams) {
+    return Stream.of(streams)
+        .map(url -> url.replaceFirst("^(mjpe?g|ip|usb):", ""))
+        .toArray(String[]::new);
+  }
+
+  public static CameraServerSource forName(String name) {
+    return sources.computeIfAbsent(name, CameraServerSource::new);
+  }
+
+  @Override
+  public SourceType getType() {
+    return CameraServerSourceType.INSTANCE;
+  }
+
+  /**
+   * Continuously reads frames from the MJPEG stream as long as the calling thread has not been interrupted. Because
+   * this is a (nearly) infinite loop, this <strong>cannot</strong> be called from the JavaFX application thread. We
+   * call this in an executor service to run asynchronously.
+   */
+  private void grabForever() {
+    if (Platform.isFxApplicationThread()) {
+      throw new IllegalStateException("This may not run on the FX application thread!");
+    }
+    final Thread thread = Thread.currentThread();
```

Comment:
Imo this is clearer if we do not cache the thread.  If we are going to cache this, we should rename the variable to be currentThread.